src/contracts/solver/SolverBase.sol

```diff
@@ -38,27 +40,31 @@ contract SolverBase {
         payBids(bids)
         returns (bool success, bytes memory data)
     {
+        // stored to allow solvers to call donateToBundler on correct execution environment
+        _executionEnv = msg.sender;
+
         (success, data) = address(this).call{value: msg.value}(solverOpData);
 
         require(success, "CALL UNSUCCESSFUL");
     }
 
+    // Helper for solvers to donate ETH to the bundler via the Execution Environment
+    // As long as the safetyFirst modifier is used while calling this function,
+    // we know that msg.sender is a real Atlas Execution Environment
+    function _donateToBundler(uint256 ethToDonate, address surplusRecipient) internal {
+        IExecutionEnvironment(_executionEnv).donateToBundler{value: ethToDonate}(surplusRecipient);
+    }
+
     modifier safetyFirst(address sender) {
-        // Safety checks
         require(sender == _owner, "INVALID CALLER");
-        // uint256 msgValueOwed = msg.value;
-
         _;
-
         // NOTE: Because this is nested inside of an Atlas meta transaction, if someone is attempting
         // to innappropriately access your smart contract then THEY will have to pay for the gas...
         // so feel free to run the safety checks at the end of the call.
         // NOTE: The solverSafetyCallback is mandatory - if it is not called then the solver
         // transaction will revert.  It is payable and can be used to repay a msg.value loan from the
         // Atlas Escrow.
-        
-        // TODO review - this has been replaced by the repayBorrowedEth modifier
-        // require(ISafetyLocks(_escrow).solverSafetyCallback{value: msgValueOwed}(msg.sender), "INVALID SEQUENCE");
+        require(ISafetyLocks(_escrow).solverSafetyCallback(msg.sender), "SOLVER SAFETY CHECK FAILED");
```

Comment:
Could we have the donateToBundler func double as the solverSafetyCallback - that way they just need to do one or the other, not both?

Reply:
Yeah that makes sense, we could merge them into 1 function, with an optional donation param. Will implement that.

Reply:
Let's hold off on this until we progress further with atlETH, as the two will probably be intertwined for gas efficiency.